timemachine/md/builders.py

```diff
@@ -1,25 +1,90 @@
 import os
-from typing import Union
+from typing import List, Optional, Union
 
 import numpy as np
-from openmm import Vec3, app, unit
+from numpy.typing import NDArray
+from openmm import app, unit
+from rdkit import Chem
 
+from timemachine.fe.utils import get_romol_conf
 from timemachine.ff import sanitize_water_ff
+from timemachine.potentials.jax_utils import idxs_within_cutoff
 
 
-def strip_units(coords):
+def strip_units(coords) -> NDArray[np.float64]:
     return np.array(coords.value_in_unit_system(unit.md_unit_system))
 
 
-def build_protein_system(host_pdbfile: Union[app.PDBFile, str], protein_ff: str, water_ff: str):
+def remove_clashy_waters(
+    modeller: app.Modeller,
+    host_coords: NDArray[np.float64],
+    box: NDArray[np.float64],
+    water_idxs: NDArray,
+    mols: List[Chem.Mol],
+    clash_distance: float = 0.4,
+):
+    """Remove waters from an OpenMM modeler that clash with a set of molecules
+
+    Parameters
+    ----------
+    modeller: app.Modeller
+        Modeller to update in place
+
+    host_coords: NDArray[np.float64]
+        Coordinates of host, may be different than modeller.positions
+
+    box: NDArray[np.float64]
+        Box to evaluate PBCs under
+
+    water_idxs: NDArray[int]
+        The water idxs to consider for deletion
+
+    mols: List[Mol]
+        List of molecules to determine which waters are clashy
+
+    clash_distance: float
+        Distance from a ligand atom to a water atom to consider as a clash, in nanometers
+    """
+    water_coords = host_coords[water_idxs]
+    ligand_coords = np.concatenate([get_romol_conf(mol) for mol in mols])
+    clashy_idxs = idxs_within_cutoff(water_coords, ligand_coords, box, cutoff=clash_distance)
+    if len(clashy_idxs) == 0:
+        return
+    # Offset the clashy idxs with the first atom idx, else could be pointing at non-water atoms
+    clashy_idxs += np.min(water_idxs)
+    all_atoms = list(modeller.topology.atoms())
+    waters_to_delete = set()
+    for idx in clashy_idxs:
+        atom = all_atoms[idx]
+        waters_to_delete.add(atom.residue)
+        assert atom.residue.name == "HOH"
+    modeller.delete(list(waters_to_delete))
```

Comment:
I'm slightly confused by the use of `set` here. Is the implication that multiple indices in `clashy_idxs` might refer to the same residue?

Reply:
If you think of a water molecule that is completely overlapping with a ligand, that will be three atoms that clash which all point to the same residue. Maybe the modeller handles duplicates, but seemed cleanest to deduplicate up front

Reply:
Makes sense. Was just confused and thinking the list was at the level of individual atoms vs. water molecules.